### PR TITLE
Add Node.js code sample for Cloud Functions + Memorystore for Redis

### PR DESCRIPTION
**Pull request by ericschmidtatwork**
I adapted the existing sample from https://cloud.google.com/memorystore/docs/redis/connect-redis-instance-flex.

### CODE REVIEW

**Comment by fhinkel** [comment]
Thanks! Can you add test cases please?

**Comment by ericschmidtatwork** [comment]
@ace-n If possible, do you mind adding tests? The PR is editable. Otherwise, it'll take me a bit to add them because I'm not familiar with the testing setup.

**Comment by ace-n** [comment]
I can if this is a) urgent or b) a one-off.

If you plan on writing Node samples regularly, it's probably better if
you're familiar with the testing setup.

If that's the case, feel free to ping me and/or put something on my
calendar if you have questions.

File: functions/memorystore/redis/index.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+const REDISPORT = process.env.REDISPORT || 6379;
+
+const redisClient = redis.createClient(REDISPORT, REDISHOST);
+redisClient.on('error', err => console.error('ERR:REDIS:', err));
+
+exports.visitCount = (req, res) => {
+  redisClient.incr('visits', (err, reply) => {
+    if (err) {
+      console.log(err);
+      res.status(500).send(err.message);
```
Comment (grant) [inline functions/memorystore/redis/index.js]: You could:
```js
return res.status(500).send(err.message);
```

**Comment by ericschmidtatwork** [comment]
@ace-n I took a stab at adding test cases, let me know what you think!

File: functions/memorystore/redis/test/index.test.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ * limitations under the License.
+ */
+
+'use strict';
+
+const proxyquire = require('proxyquire').noCallThru();
+const sinon = require('sinon');
+const assert = require('assert');
+
+function getSample({error=false}) {
```
Comment (ace-n) [inline functions/memorystore/redis/test/index.test.js]: Can we have these connect to an actual Redis instance (perhaps using the `nodejs-docs-samples-tests` GCP project)? [I added you as an Editor - but ping me if you need more access.]

@fhinkel and I are trying to phase out unit-only tests where possible.

Comment (ericschmidtatwork) [reply functions/memorystore/redis/test/index.test.js]: I'm not sure how to achieve that - the function under test would have to be deployed on GCF to connect to an actual Memorystore instance. Would we have the tests deploy and then delete the function? :thinking:

**Comment by ericschmidtatwork** [comment]
@ace-n Just a ping :) What do you think a good approach for testing this is? Is there a public Redis instance we can use?

**Comment by ace-n** [comment]
@ericschmidtatwork followed up internally. :)

File: functions/memorystore/redis/index.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+
+'use strict';
+
+// [START functions_memorystore_redis]
+
+const redis = require('redis');
+
+const REDISHOST = process.env.REDISHOST || 'localhost';
```
Comment (ace-n) [inline functions/memorystore/redis/index.js]: Nit: can use use underscores here? i.e. `REDIS_HOST` and `REDIS_PORT`.

File: functions/memorystore/redis/index.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+const REDISPORT = process.env.REDISPORT || 6379;
+
+const redisClient = redis.createClient(REDISPORT, REDISHOST);
+redisClient.on('error', err => console.error('ERR:REDIS:', err));
+
+exports.visitCount = (req, res) => {
+  redisClient.incr('visits', (err, reply) => {
```
Comment (ace-n) [inline functions/memorystore/redis/index.js]: Can we convert this to promises directly (if `incrAsync` is a supported function), or use [`util.promisify`](https://nodejs.org/dist/latest-v8.x/docs/api/util.html) here?

File: functions/memorystore/redis/package.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  "dependencies": {
+    "redis": "^2.8.0"
+  },
+  "devDependencies": {
+    "@google-cloud/nodejs-repo-tools": "^3.3.0",
+    "mocha": "^6.1.4",
+    "proxyquire": "^2.1.0",
+    "sinon": "^7.3.2"
+  },
+  "cloud-repo-tools": {
```
Comment (ace-n) [inline functions/memorystore/redis/package.json]: @fhinkel is this still relevant if we plan to nix `nodejs-repo-tools`?

(It's useful documentation of what env vars we need - but perhaps there's a better place to pit that?)

File: functions/memorystore/redis/test/index.test.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      setTimeout(resolve, 1500);
+    });
+
+    assert.strictEqual(mocks.res.status.called, false);
+    assert.strictEqual(mocks.res.send.called, false);
+    assert.strictEqual(mocks.res.writeHead.calledWith(200), true);
+    assert.strictEqual(mocks.res.end.calledOnce, true);
+
+    const response = mocks.res.end.firstCall.args[0];
+    assert.strictEqual(response.startsWith('Visit count:'), true);
```
Comment (ace-n) [inline functions/memorystore/redis/test/index.test.js]: @fhinkel should we use `assert.ok(...)` (or even plain old `assert(...)`) here instead?

File: functions/memorystore/redis/package.json (hunk 5 above)
Comment (fhinkel) [reply functions/memorystore/redis/package.json]: If you don't use `repo-tools` in the tests, you can delete that dependency and the `cloud-repo-tool` field. Thanks!

File: functions/memorystore/redis/index.js (hunk 3 above)
Comment (ericschmidtatwork) [reply functions/memorystore/redis/index.js]: All the existing samples use `REDISHOST` and `REDISPORT`, so while I agree with you, I think it's better to be consistent, which would also help with writing the docs.

File: functions/memorystore/redis/index.js (hunk 4 above)
Comment (ericschmidtatwork) [reply functions/memorystore/redis/index.js]: Done :+1:

File: functions/memorystore/redis/package.json (hunk 5 above)
Comment (ericschmidtatwork) [reply functions/memorystore/redis/package.json]: Done :+1:

File: functions/memorystore/redis/test/index.test.js (hunk 6 above)
Comment (fhinkel) [reply functions/memorystore/redis/test/index.test.js]: I prefer `assert`, it's shorter.

Comment (ericschmidtatwork) [reply functions/memorystore/redis/test/index.test.js]: Done

**Comment by fhinkel** [comment]
@ace-n Any more review comments or can we merge this? Thanks

**Comment by ericschmidtatwork** [comment]
Much of the value of the README is linking to the tutorial for this code sample, which hasn't been published yet. I plan to add a link once the tutorial is published. If necessary I can add a bare README now and update it with a link later.

**Comment by grant** [comment]
Let's at least add a bare README and update it with links to external sites later.

**Comment by fhinkel** [comment]
I'll land this, hoping all tests will pass 🙏

**Comment by ericschmidtatwork** [comment]
Wonderful! Thanks!

**Pull request merged**